### PR TITLE
feat: set-root script

### PR DESCRIPTION
**Pull request by dohaki**
Adds set-root script that can be used for testing

### CODE REVIEW

File: scripts/set-root.ts [hunk 1]
```diff
@@ -1,17 +1,102 @@
...
+  const waitBlocks = Number(options.wait);
+
+  console.log("\n2. Checking recipient amounts and duplicates...");
+  checkRecipientAmountsAndDuplicates(
+    validInputFile.recipientsWithProofs,
+    validInputFile.rewardsToDeposit
+  );
+
+  console.log("\n3. Checking reward token contract...");
+  const { ownerWallet, infuraApiKey } = assertAndGetOwnerWallet();
```
Comment (amateima) [inline scripts/set-root.ts:62]: Owner wallet should be the owner of the MerkleDistributor, RewardToken or both of them?

File: scripts/set-root.ts [hunk 2]
```diff
@@ -1,17 +1,102 @@
...
+    validInputFile,
+    waitBlocks
+  );
+
+  console.log("\n4. Checking merkle distributor contract...");
+  const merkleDistributor = MerkleDistributor__factory.connect(
+    merkleDistributorAddress,
+    ownerWallet.connect(provider)
+  );
+  await checkMerkleDistributor(merkleDistributor, validInputFile);
```
Comment (amateima) [inline scripts/set-root.ts:84]: Nice check 👍

File: scripts/set-root.ts (hunk 1 above)
Comment (amateima) [reply scripts/set-root.ts:62]: Ok. Got it. No need to rename

**Review by amateima — APPROVED** [review]

File: src/validation.ts [hunk 3]
```diff
@@ -23,6 +29,65 @@ export function assertAndGetScraperEnvVars() {
...
+  }
+}
+
+export async function checkMerkleDistributor(
+  merkleDistributor: MerkleDistributor,
+  publishedTreeFile: PublishedTreeFile
+) {
+  const nextWindowIndex = await merkleDistributor.nextCreatedIndex();
+  assert(
+    nextWindowIndex.eq(publishedTreeFile.windowIndex),
```
Comment (nicholaspai) [inline src/validation.ts:78]: Great check

File: src/validation.ts [hunk 4]
```diff
@@ -23,6 +29,65 @@ export function assertAndGetScraperEnvVars() {
...
+    `On-chain windowIndex '${nextWindowIndex}' must equal windowIndex from input file '${publishedTreeFile.windowIndex}'`
+  );
+
+  // checks if transaction would fail
+  await merkleDistributor.callStatic.setWindow(
```
Comment (nicholaspai) [inline src/validation.ts:83]: +1

**Review by nicholaspai — COMMENTED** [review]
Left one comment, i love all the tests and checks and the formatting checks with `yup`

File: scripts/set-root.ts [hunk 5]
```diff
@@ -1,17 +1,102 @@
...
+    validInputFile.recipientsWithProofs,
+    validInputFile.rewardsToDeposit
+  );
+
+  console.log("\n3. Checking reward token contract...");
+  const { ownerWallet, infuraApiKey } = assertAndGetOwnerWallet();
+  const provider = new providers.InfuraProvider(
+    validInputFile.chainId,
+    infuraApiKey
+  );
```
Comment (mrice32) [inline scripts/set-root.ts:66]: nit: prefer using a [StaticJsonRpcProvider](https://docs.ethers.io/v5/api/providers/jsonrpc-provider/#StaticJsonRpcProvider) to allow the runner to use any node url (including a local fork for testing!).

File: src/validation.ts [hunk 6]
```diff
@@ -23,6 +29,65 @@ export function assertAndGetScraperEnvVars() {
     scraperApiJwt: process.env.SCRAPER_API_JWT
   };
 }
+
+export function assertAndGetOwnerWallet() {
+  assert(process.env.INFURA_API_KEY, "Env var INFURA_API_KEY missing");
```
Comment (mrice32) [inline src/validation.ts:34]: Similar to above, prefer generalized node urls over api keys for a specific provider.

File: src/validation.ts [hunk 7]
```diff
@@ -23,6 +29,65 @@ export function assertAndGetScraperEnvVars() {
...
+    process.env.OWNER_PK || process.env.OWNER_MNEMONIC,
+    "Either OWNER_PK or OWNER_MNEMONIC need to be set"
+  );
+
+  return {
+    infuraApiKey: process.env.INFURA_API_KEY,
+    ownerWallet: process.env.OWNER_PK
+      ? new Wallet(process.env.OWNER_PK)
+      : Wallet.fromMnemonic(process.env.OWNER_MNEMONIC || "")
+  };
```
Comment (mrice32) [inline src/validation.ts:45]: Any reason not to initialize the provider and signer objects here? Right now, it seems we initialize the Wallet, but we require the caller to build the provider and signers from these building blocks.

File: src/schemas.ts [hunk 8]
```diff
@@ -1,5 +1,6 @@
 import * as yup from "yup";
 import { BigNumber, constants, utils } from "ethers";
+import * as cid from "multiformats";
```
Comment (mrice32) [inline src/schemas.ts:3]: nit: might be simpler to import this way (suggested in their docs):
```ts
import { CID } from 'multiformats/cid'
```

File: scripts/set-root.ts [hunk 9]
```diff
@@ -1,17 +1,102 @@
...
+  const options = program.opts();
+
+  console.log(`Setting root/window for ${options.input}...`);
+
+  const rawInputFile = parseInputFile(options.input);
+  const inputFile = {
+    ...rawInputFile,
+    // We transform this to an array because yup doesn't have good support for
+    // schemas with dynamic keys and type inference
+    recipientsWithProofs: Object.values(rawInputFile.recipientsWithProofs)
```
Comment (mrice32) [inline scripts/set-root.ts:43]: Sorry, just so I understand, what are the keys of rawInputFile.recipientsWithProofs?

**Review by mrice32 — COMMENTED** [review]
Looks awesome! Super minor comments -- looks great overall

File: scripts/set-root.ts (hunk 9 above)
Comment (dohaki) [reply scripts/set-root.ts:43]: The keys are the addresses of a recipient currently. See https://github.com/across-protocol/sdk-v2/blob/10d33e4bea3c87ec4fabd0793e6edbd1c5742d48/src/merkleDistributor/model/Distribution.ts#L1

File: src/validation.ts [hunk 10]
```diff
@@ -23,6 +29,65 @@ export function assertAndGetScraperEnvVars() {
...
+  expandedERC20: ExpandedERC20,
+  merkleDistributorAddress: string,
+  publishedTreeFile: PublishedTreeFile,
+  wait = 1
+) {
+  const currentAllowance = await expandedERC20.allowance(
+    await expandedERC20.signer.getAddress(),
+    merkleDistributorAddress
+  );
+  if (currentAllowance.lt(publishedTreeFile.rewardsToDeposit)) {
```
[the remark this answers is not in the document]
Comment (dohaki) [reply src/validation.ts:65]: Yes, makes sense! Changed [here](https://github.com/across-protocol/merkle-distributor/pull/3/commits/03120899f65a4726ad29a65d8690c26964370215)

File: scripts/set-root.ts (hunk 5 above)
Comment (dohaki) [reply scripts/set-root.ts]: Agree, changed [here](https://github.com/across-protocol/merkle-distributor/pull/3/commits/03120899f65a4726ad29a65d8690c26964370215)

File: src/validation.ts (hunk 7 above)
Comment (dohaki) [reply src/validation.ts]: Yea no reason. I connected the provider directly to the wallet [here](https://github.com/across-protocol/merkle-distributor/pull/3/commits/03120899f65a4726ad29a65d8690c26964370215)

File: scripts/set-root.ts (hunk 9 above)
Comment (mrice32) [reply scripts/set-root.ts:43]: Ah I see, so the keys are redundant with the contents (recipient is in both places), so it's okay to just drop the keys. Makes sense!

**Pull request merged**